src/NuGet.Services.AzureSearch/Owners2AzureSearch/Owners2AzureSearchCommand.cs

```diff
@@ -0,0 +1,106 @@
+            _ownerDataClient = ownerDataClient ?? throw new ArgumentNullException(nameof(ownerDataClient));
+            _ownerSetComparer = ownerSetComparer ?? throw new ArgumentNullException(nameof(ownerSetComparer));
+            _indexActionBuilder = indexActionBuilder ?? throw new ArgumentNullException(nameof(indexActionBuilder));
+            _batchPusherFactory = batchPusherFactory ?? throw new ArgumentNullException(nameof(batchPusherFactory));
+            _options = options ?? throw new ArgumentNullException(nameof(options));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+
+            if (_options.Value.MaxConcurrentBatches <= 0)
+            {
+                throw new ArgumentException(
+                    $"The {nameof(AzureSearchJobConfiguration.MaxConcurrentBatches)} must be greater than zero.",
+                    nameof(options));
+            }
+        }
+
+        public async Task ExecuteAsync()
+        {
+            _logger.LogInformation("Fetching old owner data from blob storage.");
+            var storageResult = await _ownerDataClient.ReadLatestIndexedAsync();
+
+            _logger.LogInformation("Fetching new owner data from the database.");
+            var databaseResult = await _databaseOwnerFetcher.GetPackageIdToOwnersAsync();
+
+            _logger.LogInformation("Detecting owner changes.");
+            var changes = _ownerSetComparer.Compare(storageResult.Result, databaseResult);
+            var changesBag = new ConcurrentBag<IdAndValue<string[]>>(changes.Select(x => new IdAndValue<string[]>(x.Key, x.Value)));
+            _logger.LogInformation("{Count} package IDs have owner changes.", changesBag.Count);
+
+            if (!changes.Any())
+            {
+                return;
+            }
+
+            _logger.LogInformation(
+                "Starting {Count} workers pushing owners changes to Azure Search.",
+                _options.Value.MaxConcurrentBatches);
+            var workerTasks = Enumerable
+                .Range(0, _options.Value.MaxConcurrentBatches)
+                .Select(x => WorkAsync(changesBag))
+                .ToList();
+            await Task.WhenAll(workerTasks);
+            _logger.LogInformation("All of the owner changes have been pushed to Azure Search.");
+
+            // Persist in storage the list of all package IDs that have owner changes. This allows debugging and future
+            // analytics on frequency of ownership changes.
+            _logger.LogInformation("Uploading the package IDs that have owner changes to blob storage.");
+            await _ownerDataClient.UploadChangeHistoryAsync(changes.Keys.ToList());
+
+            _logger.LogInformation("Uploading the new owner data to blob storage.");
+            await _ownerDataClient.ReplaceLatestIndexedAsync(databaseResult, storageResult.AccessCondition);
+        }
+
+        private async Task WorkAsync(ConcurrentBag<IdAndValue<string[]>> changesBag)
```

Comment:
I think this would be a useful thing to share--a method that grabs items from a `ConcurrentBag` and then processes items until none are left. I know I wrote something that did this when I worked on the monitoring jobs. Do we have anything like this today? Could we move this to a shared place if not?

Reply:
Spoke offline. This not easy to make generic and readable.